adapter/service/mediaplayer/service.go

```diff
@@ -0,0 +1,320 @@
+package mediaplayer
+
+import (
+	"fmt"
+	"slices"
+	"sync"
+
+	"github.com/futurehomeno/fimpgo"
+	"github.com/futurehomeno/fimpgo/fimptype"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/futurehomeno/cliffhanger/adapter"
+	"github.com/futurehomeno/cliffhanger/adapter/cache"
+)
+
+const (
+	// PlaybackPlay is the command to play the media.
+	PlaybackPlay PlaybackAction = "play"
+	// PlaybackPause is the command to pause the media.
+	PlaybackPause PlaybackAction = "pause"
+	// PlaybackTogglePlayPause is the command to toggle play/pause the media.
+	PlaybackTogglePlayPause PlaybackAction = "toggle_play_pause"
+	// PlaybackNextTrack is the command to play the next track.
+	PlaybackNextTrack PlaybackAction = "next_track"
+	// PlaybackPreviousTrack is the command to play the previous track.
+	PlaybackPreviousTrack PlaybackAction = "previous_track"
+)
+
+var (
+	// DefaultReportingStrategy is the default reporting strategy used by the service for periodic reports.
+	DefaultReportingStrategy = cache.ReportOnChangeOnly()
+
+	allowedActions = []PlaybackAction{PlaybackPlay, PlaybackPause, PlaybackTogglePlayPause, PlaybackNextTrack, PlaybackPreviousTrack}
+)
+
+// Controller is an interface representing an actual device.
+type Controller interface {
+	// SetPlayback sets the playback state.
+	SetPlayback(action PlaybackAction) error
+	// Playback returns the playback report.
+	Playback() (PlaybackAction, error)
+	// SetPlaybackMode sets the playback mode.
+	SetPlaybackMode(mode PlaybackMode) error
+	// PlaybackMode returns the playback mode.
+	PlaybackMode() (PlaybackMode, error)
+	// SetVolume sets the volume level.
+	SetVolume(level int64) error
+	// Volume returns the volume level.
+	Volume() (int64, error)
+	// SetMute sets the mute state.
+	SetMute(mute bool) error
+	// Mute returns the mute state.
+	Mute() (bool, error)
+	// Metadata returns the metadata.
+	Metadata() (Metadata, error)
+}
+
+// Service is an interface representing a mediaplayer FIMP service.
+type Service interface {
+	adapter.Service
+
+	SetPlayback(action PlaybackAction) error
+	SendPlaybackReport(force bool) (bool, error)
+	SetPlaybackMode(mode PlaybackMode) error
+	SendPlaybackModeReport(force bool) (bool, error)
+	SetVolume(level int64) error
+	SendVolumeReport(force bool) (bool, error)
+	SetMute(mute bool) error
+	SendMuteReport(force bool) (bool, error)
+	SendMetadataReport(force bool) (bool, error)
+}
+
+type (
+	// PlaybackAction represents a playback action.
+	PlaybackAction string
+	// PlaybackMode represents a playback mode.
+	PlaybackMode struct {
```

Comment:
The list of playback modes should be dynamic and based on the service specification. Which also means that we should not accept or report modes the service does not support.